# Практикум: предобработка данных в pandas

**Датасет:** `credit_risk_dataset.csv` — заявки на кредит и факт просрочки (`loan_status`).

**Цель ноутбука** — пройти путь от «сырого» CSV до таблицы, которой можно доверять: загрузка → типы → очистка → пропуски → выбросы → статистика → профилирование → гипотезы.

Главное правило (из практикума): **сначала смотрим, потом чистим** — нельзя чистить то, чего ещё не видел.

Используем только `numpy` и `pandas`.

| № | Этап | Ключевые инструменты |
|---|------|---------------------|
| 1 | Загрузка и первичный осмотр | `read_csv(nrows=…)`, `info`, `describe`, `value_counts` |
| 2 | Преобразование типов | `astype('category')`, `pd.Categorical(ordered=True)`, `map` |
| 3 | Очистка: дубликаты и невалидные значения | `duplicated`, `drop_duplicates`, `mask` |
| 4 | Заполнение пропусков | `isna`, `fillna`, `groupby().transform('median')` |
| 5 | Выбросы | `quantile`, IQR, `np.log`, `groupby` |
| 6 | Итог очистки | сравнение «до / после» |
| 7 | Описательная статистика | `describe`, `skew`, `kurtosis`, `groupby().agg` |
| 8 | Профилирование | свой «паспорт» столбцов, `groupby`, `qcut`, `pivot_table` |
| 9 | Гипотезы (не более 3) | — |

In [1]:
import numpy as np
import pandas as pd

from os.path import join
from typing import Any, Dict, List, Tuple
from IPython.display import display, Markdown

pd.set_option('display.max_columns', 50)

print('numpy :', np.__version__)
print('pandas:', pd.__version__)

numpy : 2.0.2
pandas: 2.3.3


## Описание данных

| Название признака | Описание | Категория |
|--------------|-------------|-------------|
| person_age | Возраст | Атрибут заёмщика |
| person_income | Годовой доход | Атрибут заёмщика |
| person_home_ownership | Владение жильём | Атрибут заёмщика |
| person_emp_length | Стаж работы (в годах) | Атрибут заёмщика |
| loan_intent | Цель кредита | Информация о кредите |
| loan_grade | Кредитная категория (A — лучшая, G — худшая) | Информация о кредите |
| loan_amnt | Сумма кредита | Информация о кредите |
| loan_int_rate | Процентная ставка | Информация о кредите |
| loan_status | Статус кредита (0 — нет просрочки, 1 — просрочка) | **Целевая переменная** |
| loan_percent_income | Доля суммы кредита от годового дохода | Информация о кредите |
| cb_person_default_on_file | Историческая просрочка (Y/N) | Кредитная история |
| cb_person_cred_hist_length | Длина кредитной истории (в годах) | Кредитная история |

In [2]:
DATA_PATH = join('data', 'credit_risk_dataset.csv')

TARGET = 'loan_status'

NUMERIC_COLS = [
    'person_age', 'person_income', 'person_emp_length',
    'loan_amnt', 'loan_int_rate', 'loan_percent_income',
    'cb_person_cred_hist_length',
]
CATEGORICAL_COLS = [
    'person_home_ownership', 'loan_intent',
    'loan_grade', 'cb_person_default_on_file',
]

## 1. Загрузка данных

### 1.1. Сначала заглядываем в файл

Типичные «грабли» `read_csv`: неверная кодировка, разделитель `;` вместо `,`, десятичная запятая. Поэтому сначала смотрим на несколько «сырых» строк файла и загружаем только первые строки через `nrows`.

In [3]:
with open(DATA_PATH, encoding='utf-8') as f:
    for _ in range(3):
        print(repr(f.readline()))

'person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length\n'
'22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3\n'
'21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.1,N,2\n'


In [4]:
pd.read_csv(DATA_PATH, nrows=5)

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,D,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,B,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,C,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,C,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,C,35000,14.27,1,0.55,Y,4


**Что видно:** файл читается как UTF-8, разделитель — запятая, дробная часть — через точку (`16.02`), первая строка — заголовок. Значит, параметры `read_csv` по умолчанию подходят, и `sep`, `decimal`, `encoding` указывать не нужно.

### 1.2. Загружаем весь файл

Сохраним «сырую» копию `df_raw`, чтобы в конце сравнить данные **до** и **после** обработки.

In [5]:
df_raw = pd.read_csv(DATA_PATH)
df = df_raw.copy()

print('Размер:', df.shape)

Размер: (32581, 12)


### 1.3. Первичный осмотр

Привычка после любой загрузки: смотрим `shape`, типы, число уникальных значений и пропусков.

In [6]:
overview = pd.DataFrame({
    'dtype': df.dtypes.astype(str),
    'n_unique': df.nunique(),
    'n_missing': df.isna().sum(),
    'missing_%': (df.isna().mean() * 100).round(2),
    'example': df.iloc[0],
})
overview

,dtype,n_unique,n_missing,missing_%,example
person_age,int64,58,0,0.00,22
person_income,int64,4295,0,0.00,59000
person_home_ownership,object,4,0,0.00,RENT
person_emp_length,float64,36,895,2.75,123.0
loan_intent,object,6,0,0.00,PERSONAL
loan_grade,object,7,0,0.00,D
loan_amnt,int64,753,0,0.00,35000
loan_int_rate,float64,348,3116,9.56,16.02
loan_status,int64,2,0,0.00,1
loan_percent_income,float64,77,0,0.00,0.59


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  object 
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  object 
 5   loan_grade                  32581 non-null  object 
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  object 
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), object(4)
memory usage: 3.0+ MB


In [8]:
df.describe().T.round(2)

,count,mean,std,min,25%,50%,75%,max
person_age,32581.0,27.73,6.35,20.00,23.00,26.00,30.00,144.00
person_income,32581.0,66074.85,61983.12,4000.00,38500.00,55000.00,79200.00,6000000.00
person_emp_length,31686.0,4.79,4.14,0.00,2.00,4.00,7.00,123.00
loan_amnt,32581.0,9589.37,6322.09,500.00,5000.00,8000.00,12200.00,35000.00
loan_int_rate,29465.0,11.01,3.24,5.42,7.90,10.99,13.47,23.22
loan_status,32581.0,0.22,0.41,0.00,0.00,0.00,0.00,1.00
loan_percent_income,32581.0,0.17,0.11,0.00,0.09,0.15,0.23,0.83
cb_person_cred_hist_length,32581.0,5.80,4.06,2.00,3.00,4.00,8.00,30.00


In [9]:
for col in CATEGORICAL_COLS:
    display(Markdown(f'**{col}**'))
    display(df[col].value_counts(dropna=False).to_frame('count'))

**person_home_ownership**

,count
person_home_ownership,
RENT,16446
MORTGAGE,13444
OWN,2584
OTHER,107


**loan_intent**

,count
loan_intent,
EDUCATION,6453
MEDICAL,6071
VENTURE,5719
PERSONAL,5521
DEBTCONSOLIDATION,5212
HOMEIMPROVEMENT,3605


**loan_grade**

,count
loan_grade,
A,10777
B,10451
C,6458
D,3626
E,964
F,241
G,64


**cb_person_default_on_file**

,count
cb_person_default_on_file,
N,26836
Y,5745


In [10]:
print('Полных дубликатов строк:', df.duplicated().sum())
print()
print('Баланс целевой переменной, %:')
print((df[TARGET].value_counts(normalize=True) * 100).round(2))

Полных дубликатов строк: 165

Баланс целевой переменной, %:
loan_status
0    78.18
1    21.82
Name: proportion, dtype: float64


**Что уже видно — список задач на очистку:**

1. `person_age`: максимум **144** года — невозможное значение.
2. `person_emp_length`: максимум **123** года стажа — тоже невозможно.
3. Пропуски в `person_emp_length` (~3%) и `loan_int_rate` (~10%).
4. `person_income`: среднее заметно больше медианы, а максимум — **6 млн** при медиане 55 тыс. Сильная правосторонняя асимметрия и, возможно, выбросы.
5. **165** полных дубликатов строк.
6. Текстовые столбцы хранятся как строки, хотя по смыслу это категории.
7. Целевая переменная несбалансирована: просрочек около 22%.

## 2. Преобразование типов данных

**Даты.** В датасете нет столбцов с датами, поэтому `pd.to_datetime` не нужен. (Если бы были — всегда указывали бы `format=`, иначе pandas может перепутать день и месяц.)

**Комплексные параметры.** Строк вида `"5 лет"`, `"12,5%"` или `"1 500"` нет — все числа уже распознаны как `int64` / `float64`. Парсить ничего не нужно.

**Что стоит поменять:**

- `person_home_ownership`, `loan_intent` → `category`: мало уникальных значений, экономим память;
- `loan_grade` → **упорядоченная** категория (A < B < … < G): порядок имеет смысл, и его можно использовать в сравнениях;
- `cb_person_default_on_file` → `bool`: по смыслу это «да/нет».

Перед преобразованием `Y/N` проверим, что других значений нет: `map` превратит любое неожиданное значение в `NaN`.

In [11]:
print('cb_person_default_on_file:', df['cb_person_default_on_file'].unique())
print('loan_grade               :', sorted(df['loan_grade'].unique()))

cb_person_default_on_file: ['Y' 'N']
loan_grade               : ['A', 'B', 'C', 'D', 'E', 'F', 'G']


In [12]:
GRADE_ORDER = ['A', 'B', 'C', 'D', 'E', 'F', 'G']

mem_before = df.memory_usage(deep=True).sum() / 2**20

df = df.assign(
    person_home_ownership=df['person_home_ownership'].astype('category'),
    loan_intent=df['loan_intent'].astype('category'),
    loan_grade=pd.Categorical(df['loan_grade'], categories=GRADE_ORDER, ordered=True),
    cb_person_default_on_file=df['cb_person_default_on_file'].map({'Y': True, 'N': False}),
)

# Проверка: преобразование не должно было породить новых пропусков
assert df[CATEGORICAL_COLS].isna().sum().sum() == 0

mem_after = df.memory_usage(deep=True).sum() / 2**20
print(f'Память: {mem_before:.2f} МБ -> {mem_after:.2f} МБ')
df.dtypes

Память: 9.62 МБ -> 2.11 МБ


person_age                       int64
person_income                    int64
person_home_ownership         category
person_emp_length              float64
loan_intent                   category
loan_grade                    category
loan_amnt                        int64
loan_int_rate                  float64
loan_status                      int64
loan_percent_income            float64
cb_person_default_on_file         bool
cb_person_cred_hist_length       int64
dtype: object

In [13]:
# Упорядоченная категория поддерживает сравнения
print('Кредитов с грейдом хуже C:', (df['loan_grade'] > 'C').sum())

Кредитов с грейдом хуже C: 4895


> **Совет.** Те же типы можно задать сразу при чтении файла — так данные ни на секунду не окажутся «не того» типа:
> ```python
> pd.read_csv(
>     DATA_PATH,
>     dtype={'person_home_ownership': 'category',
>            'loan_intent': 'category',
>            'loan_grade': pd.CategoricalDtype(GRADE_ORDER, ordered=True)},
>     true_values=['Y'], false_values=['N'],
> )
> ```
> В учебных целях мы сделали это отдельным шагом, чтобы увидеть проверку и выигрыш по памяти.

## 3. Очистка данных: дубликаты и невалидные значения

### 3.1. Дубликаты

Удалять дубли нужно не всегда: иногда повтор — это реальное повторное событие. Здесь в строке 12 признаков, включая точный доход и ставку. Случайное совпадение **всех** значений маловероятно, поэтому считаем такие строки техническими дублями выгрузки и удаляем.

In [14]:
dup_mask = df.duplicated(keep=False)   # keep=False — показать все копии, а не только повторы
print('Строк-дубликатов (без первых вхождений):', df.duplicated().sum())
print(df[dup_mask].isnull().sum())
df[dup_mask].sort_values(list(df.columns)).head(15)

Строк-дубликатов (без первых вхождений): 165
person_age                     0
person_income                  0
person_home_ownership          0
person_emp_length             16
loan_intent                    0
loan_grade                     0
loan_amnt                      0
loan_int_rate                 42
loan_status                    0
loan_percent_income            0
cb_person_default_on_file      0
cb_person_cred_hist_length     0
dtype: int64


,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
15944,21,8088,RENT,NaN,MEDICAL,C,1200,15.23,0,0.15,True,2
16835,21,8088,RENT,NaN,MEDICAL,C,1200,15.23,0,0.15,True,2
2431,21,15600,RENT,0.0,MEDICAL,A,2800,7.40,1,0.18,False,4
17758,21,15600,RENT,0.0,MEDICAL,A,2800,7.40,1,0.18,False,4
2498,21,18000,RENT,0.0,DEBTCONSOLIDATION,A,3000,7.90,1,0.17,False,2
17601,21,18000,RENT,0.0,DEBTCONSOLIDATION,A,3000,7.90,1,0.17,False,2
1778,21,20000,RENT,0.0,MEDICAL,A,2000,NaN,0,0.10,False,3
17642,21,20000,RENT,0.0,MEDICAL,A,2000,NaN,0,0.10,False,3
740,21,21600,OWN,NaN,VENTURE,A,7125,6.99,0,0.33,False,3
17283,21,21600,OWN,NaN,VENTURE,A,7125,6.99,0,0.33,False,3


In [15]:
rows_before = len(df)
df = df.drop_duplicates().reset_index(drop=True)
print(f'Удалено дубликатов: {rows_before - len(df)}, осталось строк: {len(df)}')

Удалено дубликатов: 165, осталось строк: 32416


### 3.2. Невалидные значения

**Невалидное значение ≠ выброс.** Выброс — редкое, но возможное значение (доход 2 млн). Невалидное — **невозможное** по смыслу (возраст 144 года). Это ошибка ввода, и правильно её заменить на пропуск (`NaN`), а затем обработать вместе с остальными пропусками.

Посмотрим на самые большие значения возраста и стажа:

In [16]:
display(Markdown('**Топ по person_age**'))
display(df.nlargest(7, 'person_age')[['person_age', 'person_emp_length', 'person_income', 'cb_person_cred_hist_length']])

display(Markdown('**Топ по person_emp_length**'))
display(df.nlargest(4, 'person_emp_length')[['person_age', 'person_emp_length', 'person_income']])

**Топ по person_age**

,person_age,person_emp_length,person_income,cb_person_cred_hist_length
81,144,4.0,250000,3
183,144,4.0,200000,2
32132,144,12.0,6000000,25
575,123,2.0,80004,3
747,123,7.0,78000,4
32251,94,1.0,24000,27
32341,84,2.0,94800,24


**Топ по person_emp_length**

,person_age,person_emp_length,person_income
0,22,123.0,59000
210,21,123.0,192000
32190,78,41.0,48000
32350,53,38.0,106000


Сформулируем **правила** на основе предметной области:

| Правило | Обоснование |
|---|---|
| `person_age > 100` | После 94 лет сразу идут 123 и 144 — явный разрыв; таких заёмщиков не бывает. |
| `person_emp_length > person_age − 14` | Нельзя работать дольше, чем живёшь: стаж 123 года у 21-летнего невозможен. 14 лет — минимальный возраст начала трудовой деятельности. |
| `cb_person_cred_hist_length > person_age` | Кредитная история не может быть длиннее жизни. |

Сначала считаем все маски на исходных данных, потом заменяем значения. `Series.mask(cond)` ставит `NaN` там, где условие истинно.

In [17]:
MAX_AGE = 100
MIN_WORK_AGE = 14

invalid_age = df['person_age'] > MAX_AGE
invalid_emp = df['person_emp_length'] > df['person_age'] - MIN_WORK_AGE
invalid_hist = df['cb_person_cred_hist_length'] > df['person_age']

pd.Series({
    'возраст > 100': invalid_age.sum(),
    'стаж > возраст - 14': invalid_emp.sum(),
    'кредитная история > возраст': invalid_hist.sum(),
}, name='нарушений')

возраст > 100                  5
стаж > возраст - 14            2
кредитная история > возраст    0
Name: нарушений, dtype: int64

In [18]:
df['person_age'] = df['person_age'].mask(invalid_age)
df['person_emp_length'] = df['person_emp_length'].mask(invalid_emp)

df[['person_age', 'person_emp_length']].isna().sum()

person_age             5
person_emp_length    889
dtype: int64

Обратите внимание: `person_age` стал `float64`. В NumPy `NaN` существует только у `float`, поэтому целочисленный столбец с пропуском превращается в дробный. После заполнения пропусков вернём целый тип.

## 4. Заполнение пропусков

### 4.1. Сколько пропусков

In [19]:
missing = pd.DataFrame({
    'пропусков': df.isna().sum(),
    'доля_%': (df.isna().mean() * 100).round(2),
})
missing[missing['пропусков'] > 0]

,пропусков,доля_%
person_age,5,0.02
person_emp_length,889,2.74
loan_int_rate,3095,9.55


### 4.2. Случайны ли пропуски?

Прежде чем заполнять, стоит проверить, **связан ли сам факт пропуска с целевой переменной**. Если да — пропуск несёт информацию, и её нельзя потерять при заполнении.


Что может стоять за пропуском стажа (person_emp_length), в случае, если **связан ли сам факт пропуска с целевой переменной**? Предположения:

- Человеку нечего указать. Безработный, студент, фрилансер, самозанятый, неформальная занятость. У таких людей обычно нестабильный доход, а это прямо связано с риском просрочки.
- Человек не захотел указывать. Например, стаж маленький или работа недавно потеряна, и заявитель предпочёл оставить поле пустым. Само умолчание — сигнал.
- Особенность процесса. Заявка пришла через канал или старую версию анкеты, где поле было необязательным. Или для небольших сумм банк стаж не спрашивал. Тогда пропуск кодирует канал или тип продукта, а они тоже могут быть связаны с риском.

In [20]:
rows = []
for col in ['person_emp_length', 'loan_int_rate']:
    is_missing = df[col].isna()
    rows.append({
        'столбец': col,
        'доля просрочек, значение есть': df.loc[~is_missing, TARGET].mean(),
        'доля просрочек, значение пропущено': df.loc[is_missing, TARGET].mean(),
    })

(pd.DataFrame(rows).set_index('столбец') * 100).round(1)

,"доля просрочек, значение есть","доля просрочек, значение пропущено"
столбец,,
person_emp_length,21.6,31.7
loan_int_rate,22.0,20.7


In [21]:
emp_missing = df['person_emp_length'].isna()
df.groupby(emp_missing).agg(
    n=(TARGET, 'size'),
    default_rate=(TARGET, 'mean'),
    median_income=('person_income', 'median'),
).rename_axis('стаж пропущен').round(3)

,n,default_rate,median_income
стаж пропущен,,,
False,31527,0.216,56000.0
True,889,0.317,36000.0


- `loan_int_rate`: доля просрочек почти не зависит от того, есть ставка или нет. Похоже, пропуск случаен — просто заполняем.
- `person_emp_length`: среди заёмщиков **без указанного стажа** просрочек около 32% против ~22% у остальных, а их медианный доход заметно ниже.

> **Гипотеза 1. (H1)** Отсутствие стажа работы в заявке — самостоятельный сигнал риска: вероятно, это безработные или люди с нестабильной занятостью. Поэтому сохраняем признак-флаг `emp_length_missing` **до** заполнения пропусков.

In [22]:
df['emp_length_missing'] = df['person_emp_length'].isna()
df['emp_length_missing'].sum()

np.int64(889)

### 4.3. Среднее или медиана?

Если распределение асимметрично или есть выбросы, среднее «перетягивается» хвостом. В таком случае медиана надёжнее.

In [23]:
cols_with_na = ['person_age', 'person_emp_length', 'loan_int_rate']
pd.DataFrame({
    'среднее': df[cols_with_na].mean(),
    'медиана': df[cols_with_na].median(),
    'асимметрия (skew)': df[cols_with_na].skew(),
}).round(2)

,среднее,медиана,асимметрия (skew)
person_age,27.73,26.00,1.94
person_emp_length,4.78,4.00,1.25
loan_int_rate,11.02,10.99,0.21


У возраста и стажа есть правосторонняя асимметрия (skew > 1), поэтому для них используем медиану.

Процентная ставка **сильно зависит от грейда** (чем хуже грейд — тем выше ставка). Если заполнить её одной общей медианой, заёмщик с грейдом A и заёмщик с грейдом G получат одинаковую ставку. Правильнее заполнить **медианой внутри грейда**.

In [24]:
df.groupby('loan_grade', observed=True)['loan_int_rate'].agg(['median', 'count']).round(2)

,median,count
loan_grade,,
A,7.49,9713
B,10.99,9337
C,13.48,5809
D,15.31,3309
E,16.82,880
F,18.54,214
G,20.16,59


### 4.4. Заполняем

`groupby(...).transform('median')` возвращает Series **той же длины**, что и таблица: для каждой строки — медиану её группы. Её и передаём в `fillna`.

In [25]:
grade_median_rate = df.groupby('loan_grade', observed=True)['loan_int_rate'].transform('median')
df['loan_int_rate'] = df['loan_int_rate'].fillna(grade_median_rate)

df['person_emp_length'] = df['person_emp_length'].fillna(df['person_emp_length'].median())

df['person_age'] = df['person_age'].fillna(df['person_age'].median()).round().astype('int64')

assert df.isna().sum().sum() == 0, 'Остались пропуски!'
print('Пропусков не осталось')

Пропусков не осталось


> **Важно** Медианы в этом ноутбуке посчитаны по всему датасету — для учебного анализа это нормально. При обучении модели статистики для заполнения считают **только на обучающей выборке** и затем применяют к тестовой, иначе возникает утечка данных (data leakage).

## 5. Анализ и обработка выбросов

**Выброс (outlier)** — наблюдение, которое сильно отличается от основной массы данных. В отличие от невалидного значения, выброс может быть **настоящим**: очень богатый заёмщик или очень высокая ставка.

| Тип | Описание |
| :--- | :--- |
| **Одномерный** | Аномалия видна по одному признаку (доход 6 млн при медиане 55 тыс.). |
| **Многомерный** | По каждому признаку значение нормальное, аномально только их **сочетание**. |

Поэтому выбросы **не удаляем вслепую**: сначала разбираемся, откуда они взялись.

### 5.1. Взгляд через квантили

In [26]:
df[NUMERIC_COLS].quantile([0, 0.01, 0.25, 0.5, 0.75, 0.99, 0.999, 1]).T.round(2)

,0.000,0.010,0.250,0.500,0.750,0.990,0.999,1.000
person_age,20.00,21.00,23.00,26.00,30.00,50.00,65.00,94.00
person_income,4000.00,14400.00,38542.00,55000.00,79218.00,225000.00,648000.00,6000000.00
person_emp_length,0.00,0.00,2.00,4.00,7.00,17.00,24.00,41.00
loan_amnt,500.00,1000.00,5000.00,8000.00,12250.00,29800.00,35000.00,35000.00
loan_int_rate,5.42,5.42,7.88,10.99,13.48,18.54,20.89,23.22
loan_percent_income,0.00,0.02,0.09,0.15,0.23,0.50,0.63,0.83
cb_person_cred_hist_length,2.00,2.00,3.00,4.00,8.00,17.00,29.00,30.00


### 5.2. Метод межквартильного размаха (IQR)

1. Считаем Q1 (25% квантиль) и Q3 (75% квантиль).
2. IQR = Q3 − Q1 — ширина «основной массы» данных.
3. Всё за пределами `[Q1 − 1.5·IQR; Q3 + 1.5·IQR]` — **кандидаты** в выбросы.

In [27]:
def iqr_bounds(ser: pd.Series, coef: float = 1.5) -> Tuple[float, float]:
    q1, q3 = ser.quantile([0.25, 0.75])
    iqr = q3 - q1
    return q1 - coef * iqr, q3 + coef * iqr


def iqr_report(data: pd.DataFrame, cols: List[str], coef: float = 1.5) -> pd.DataFrame:
    rows = []
    for col in cols:
        low, up = iqr_bounds(data[col], coef)
        is_low_out = (data[col] < low)
        is_up_out = (data[col] > up)
        outlier_count = is_low_out.sum() + is_up_out.sum()
        outlier_fraction = outlier_count / data[col].shape[0]
        rows.append({
            'столбец': col,
            'нижняя_граница': low,
            'верхняя_граница': up,
            'кол-во_выбросов за нижнюю границу': is_low_out.sum(),
            'кол-во_выбросов за верхнюю границу': is_up_out.sum(),
            'общее кол-во_выбросов': outlier_count,
            'общая доля_выбросов_%': outlier_fraction * 100,
        })
    return pd.DataFrame(rows).set_index('столбец').round(2)


iqr_report(df, NUMERIC_COLS)

,нижняя_граница,верхняя_граница,кол-во_выбросов за нижнюю границу,кол-во_выбросов за верхнюю границу,общее кол-во_выбросов,общая доля_выбросов_%
столбец,,,,,,
person_age,12.50,40.50,0,1486,1486,4.58
person_income,-22472.00,140232.00,0,1478,1478,4.56
person_emp_length,-5.50,14.50,0,850,850,2.62
loan_amnt,-5875.00,23125.00,0,1679,1679,5.18
loan_int_rate,-0.52,21.88,0,6,6,0.02
loan_percent_income,-0.12,0.44,0,650,650,2.01
cb_person_cred_hist_length,-4.50,15.50,0,1139,1139,3.51


**Интерпретация:**

- person_age - Граница узкая, потому что основная масса заёмщиков 23–30 лет. Но 45 или 60 лет — нормальный возраст. Невозможные значения (123, 144) убраны, максимум теперь 94.
- person_emp_length - Стаж 15–41 год реален. Согласованность с возрастом уже проверена правилом «стаж ≤ возраст − 14».
- cb_person_cred_hist_length - Долгая история у старших заёмщиков. Проверка «история ≤ возраст» нарушений не нашла.
- loan_amnt - Максимум 35 000 совпадает с 99.9% квантилем. **Это похоже на верхний лимит продукта**, а не на аномалию.
- loan_percent_income - Доли 0.44–0.83 возможны. Именно в этом хвосте доля просрочек выше 50% (раздел 8.3), и удаление убрало бы самый ценный для модели сигнал.
- person_income - доход в реальности имеет длинный правый хвост, и правило 1.5·IQR для таких данных слишком строгое. Разберём отдельно.
- loan_int_rate - рассмотрим их


Вывод: IQR — **диагностический** инструмент, а не автоматический фильтр.

In [28]:
_, up_bound = iqr_bounds(df['loan_int_rate'], 1.5)
df[df['loan_int_rate'] > up_bound]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length,emp_length_missing
6449,23,34255,RENT,7.0,VENTURE,G,6000,22.11,1,0.18,True,3,False
13499,26,44000,RENT,4.0,HOMEIMPROVEMENT,G,12500,22.11,1,0.28,True,2,False
18707,31,22000,RENT,5.0,VENTURE,F,1400,22.06,0,0.06,True,7,False
27303,28,128000,MORTGAGE,11.0,DEBTCONSOLIDATION,G,14400,22.11,1,0.11,True,8,False
27729,28,288000,MORTGAGE,12.0,VENTURE,G,18000,22.48,1,0.06,True,9,False
29118,38,83000,RENT,3.0,HOMEIMPROVEMENT,G,22250,23.22,1,0.27,True,12,False


Эти 6 ставок 21.9–23.2% относятся к грейдам F и G и согласуются с грейдом.

In [29]:
up_loan_limit = df[df['loan_amnt'] == df['loan_amnt'].max()]
print(up_loan_limit.shape)
up_loan_limit['loan_status'].value_counts(normalize=True)

(184, 13)


loan_status
0    0.673913
1    0.326087
Name: proportion, dtype: float64

Лимит — наша догадка, а не факт. На него указывает то, что сумм выше 35 000 нет вообще, а на самом значении скопилось 184 строки (на 32 000 — всего 10). Но в описании датасета об этом ничего не сказано.

### 5.3. Доход: логарифмическая шкала и явная аномалия

Для скошенных «денежных» признаков IQR разумнее считать на логарифме: `np.log` сжимает длинный хвост.

In [30]:
df['person_income_log'] = np.log(df['person_income'])

iqr_report(df, ['person_income', 'person_income_log'])

,нижняя_граница,верхняя_граница,кол-во_выбросов за нижнюю границу,кол-во_выбросов за верхнюю границу,общее кол-во_выбросов,общая доля_выбросов_%
столбец,,,,,,
person_income,-22472.00,140232.00,0,1478,1478,4.56
person_income_log,9.48,12.36,268,303,571,1.76


In [31]:
df['person_income'].nlargest(8).to_frame()

,person_income
32132,6000000
29890,2039784
32381,1900000
32332,1782000
31765,1440000
31763,1362000
17711,1200000
28960,1200000


Один доход — **6 млн** — втрое больше следующего (≈2 млн). Посмотрим на эту запись в исходных данных:

In [32]:
INCOME_LIMIT = 5_000_000
df_raw.loc[df_raw['person_income'] > INCOME_LIMIT]

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_grade,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
32297,144,6000000,MORTGAGE,12.0,PERSONAL,C,5000,12.73,0,0.0,N,25


В этой записи сразу **две аномалии**: доход 6 млн и возраст 144 года (который мы уже признали невалидным), а `loan_percent_income` = 0.00. Такой записи доверять нельзя — **удаляем** её.

Остальные крупные доходы вполне реальны. Удалять их не будем, но **пометим флагом** (всё, что выше 99.9% квантиля), чтобы при необходимости легко исключить или проанализировать отдельно.

In [33]:
rows_before = len(df)
df = df.loc[df['person_income'] <= INCOME_LIMIT].reset_index(drop=True)
print('Удалено строк:', rows_before - len(df))

income_q999 = df['person_income'].quantile(0.999)
df['is_income_outlier'] = df['person_income'] > income_q999
print(f'Помечено флагом is_income_outlier (доход > {income_q999:,.0f}):', df['is_income_outlier'].sum())

Удалено строк: 1
Помечено флагом is_income_outlier (доход > 643,032): 33


### 5.4. Многомерный выброс: ставка против грейда

По отдельности ставка 6% — нормальное значение (она лежит внутри диапазона 5.4–23.2%). Но посмотрим на ставку **внутри каждого грейда**:

In [34]:
df.groupby('loan_grade', observed=True)['loan_int_rate'].agg(['min', 'median', 'max']).round(2)

,min,median,max
loan_grade,,,
A,5.42,7.49,9.63
B,6.00,10.99,12.69
C,6.00,13.48,16.11
D,6.00,15.31,18.49
E,6.00,16.82,20.69
F,15.01,18.54,22.06
G,17.34,20.16,23.22


У грейдов B–E минимум — ровно **6.00**, хотя медиана растёт с 11% до 17%. Ставка ниже медианы лучшего грейда A для грейдов C и хуже противоречит логике кредитования: рискованный заёмщик не может получить ставку лучше надёжного. Это и есть **многомерный выброс**: ни ставка, ни грейд по отдельности не аномальны, аномально их сочетание.

Грейд B не трогаем: его диапазон ставок пересекается с диапазоном соседнего грейда A (до 9.63%), поэтому 6% для B правдоподобны.

In [35]:
grade_a_median = df.loc[df['loan_grade'] == 'A', 'loan_int_rate'].median()
suspicious_rate = (df['loan_grade'] >= 'C') & (df['loan_int_rate'] < grade_a_median)

print(f'Медиана ставки для грейда A: {grade_a_median:.2f}')
df.loc[suspicious_rate, ['loan_grade', 'loan_int_rate', 'loan_status']]

Медиана ставки для грейда A: 7.49


,loan_grade,loan_int_rate,loan_status
6194,C,6.0,0
9581,C,6.0,0
13542,D,6.0,1
20580,E,6.0,1
24316,D,6.0,0
32258,C,6.0,0


Таких записей всего несколько, и ошибочна в них, скорее всего, только ставка. Поэтому строки не удаляем, а заменяем ставку на медиану грейда — так же, как заполняли пропуски.

In [36]:
grade_median_rate = df.groupby('loan_grade', observed=True)['loan_int_rate'].transform('median')
df['loan_int_rate'] = df['loan_int_rate'].mask(suspicious_rate, grade_median_rate)

df.groupby('loan_grade', observed=True)['loan_int_rate'].agg(['min', 'median', 'max']).round(2)

,min,median,max
loan_grade,,,
A,5.42,7.49,9.63
B,6.00,10.99,12.69
C,10.28,13.48,16.11
D,11.86,15.31,18.49
E,13.43,16.82,20.69
F,15.01,18.54,22.06
G,17.34,20.16,23.22


## 6. Итог очистки: «до» и «после»

In [37]:
def mem_mb(data: pd.DataFrame) -> float:
    return round(data.memory_usage(deep=True).sum() / 2**20, 2)


pd.DataFrame({
    'до': [len(df_raw), df_raw.shape[1], df_raw.isna().sum().sum(), df_raw.duplicated().sum(),
           df_raw['person_age'].max(), df_raw['person_emp_length'].max(), mem_mb(df_raw)],
    'после': [len(df), df.shape[1], df.isna().sum().sum(), df.duplicated().sum(),
              df['person_age'].max(), df['person_emp_length'].max(), mem_mb(df)],
}, index=['строк', 'столбцов', 'пропусков', 'дубликатов', 'макс. возраст', 'макс. стаж', 'память, МБ'])

,до,после
строк,32581.00,32415.00
столбцов,12.00,15.00
пропусков,4011.00,0.00
дубликатов,165.00,0.00
макс. возраст,144.00,94.00
макс. стаж,123.00,41.00
"память, МБ",9.62,2.41


Что сделали:

- привели типы (категории, упорядоченный грейд, `bool`);
- удалили 165 технических дубликатов;
- заменили невалидные значения возраста и стажа на пропуски;
- заполнили пропуски медианой (ставку — медианой по грейду), сохранив флаг `emp_length_missing`;
- удалили 1 заведомо ошибочную запись и пометили экстремальные доходы флагом `is_income_outlier`;
- исправили противоречивые ставки внутри грейда.

Добавились 2 столбца-флага; удалено 166 строк из 32 581 (≈0.5%).

## 7. Описательная статистика

### 7.1. Числовые признаки

In [38]:
df[NUMERIC_COLS].describe().T.round(2)

,count,mean,std,min,25%,50%,75%,max
person_age,32415.0,27.73,6.21,20.00,23.00,26.00,30.00,94.00
person_income,32415.0,65908.58,52533.03,4000.00,38528.00,55000.00,79200.00,2039784.00
person_emp_length,32415.0,4.76,3.98,0.00,2.00,4.00,7.00,41.00
loan_amnt,32415.0,9593.99,6322.78,500.00,5000.00,8000.00,12250.00,35000.00
loan_int_rate,32415.0,11.02,3.21,5.42,7.88,10.99,13.48,23.22
loan_percent_income,32415.0,0.17,0.11,0.00,0.09,0.15,0.23,0.83
cb_person_cred_hist_length,32415.0,5.81,4.06,2.00,3.00,4.00,8.00,30.00


In [39]:
pd.DataFrame({
    'среднее': df[NUMERIC_COLS].mean(),
    'медиана': df[NUMERIC_COLS].median(),
    'среднее / медиана': df[NUMERIC_COLS].mean() / df[NUMERIC_COLS].median(),
    'асимметрия (skew)': df[NUMERIC_COLS].skew(),
    'эксцесс (kurtosis)': df[NUMERIC_COLS].kurtosis(),
}).round(2)

,среднее,медиана,среднее / медиана,асимметрия (skew),эксцесс (kurtosis)
person_age,27.73,26.00,1.07,1.94,5.87
person_income,65908.58,55000.00,1.20,9.77,225.78
person_emp_length,4.76,4.00,1.19,1.28,2.59
loan_amnt,9593.99,8000.00,1.20,1.19,1.42
loan_int_rate,11.02,10.99,1.00,0.21,-0.68
loan_percent_income,0.17,0.15,1.14,1.06,1.22
cb_person_cred_hist_length,5.81,4.00,1.45,1.66,3.70


- **skew > 0** — длинный правый хвост; чем больше, тем сильнее среднее «уезжает» вправо от медианы.
- **kurtosis > 0** — «тяжёлые хвосты»: экстремальных значений больше, чем у нормального распределения.

Доход по-прежнему самый скошенный признак — это естественное свойство доходов, а не ошибка.

### 7.2. Категориальные признаки и целевая переменная

In [40]:
df[CATEGORICAL_COLS].describe().T

,count,unique,top,freq
person_home_ownership,32415,4,RENT,16378
loan_intent,32415,6,EDUCATION,6411
loan_grade,32415,7,A,10703
cb_person_default_on_file,32415,2,False,26685


In [41]:
(df[TARGET].value_counts(normalize=True) * 100).round(2).to_frame('доля, %')

,"доля, %"
loan_status,
0,78.13
1,21.87


### 7.3. Статистика по группам

Именованная агрегация `agg(имя=(столбец, функция))` сразу даёт читаемые названия колонок.

In [42]:
df.groupby('loan_grade', observed=True).agg(
    n=('loan_amnt', 'size'),
    avg_rate=('loan_int_rate', 'mean'),
    median_amount=('loan_amnt', 'median'),
    default_rate=(TARGET, 'mean'),
).round(2)

,n,avg_rate,median_amount,default_rate
loan_grade,,,,
A,10703,7.34,7500.0,0.10
B,10387,11.00,8400.0,0.16
C,6437,13.47,8000.0,0.21
D,3620,15.36,9000.0,0.59
E,963,17.00,12000.0,0.64
F,241,18.60,15000.0,0.71
G,64,20.24,19375.0,0.98


## 8. Профилирование данных

Профилирование — это сводный **«паспорт»** каждого столбца (тип, пропуски, уникальность, основные статистики) плюс взгляд на то, как признаки связаны с целевой переменной. Готовые библиотеки строят такой отчёт автоматически, но соберём его сами средствами pandas/numpy — так понятнее, что в нём внутри.

### 8.1. Паспорт столбцов

In [43]:
def profile_dataset(data: pd.DataFrame) -> pd.DataFrame:
    rows = []
    for col in data.columns:
        s = data[col]
        row: Dict[str, Any] = {
            'столбец': col,
            'тип': str(s.dtype),
            'пропусков_%': round(s.isna().mean() * 100, 2),
            'уникальных': s.nunique(),
            'мода': s.mode().iloc[0],
            'доля_моды_%': round(s.value_counts(normalize=True).iloc[0] * 100, 2),
        }
        if pd.api.types.is_numeric_dtype(s) and not pd.api.types.is_bool_dtype(s):
            row.update({
                'min': s.min(),
                'медиана': s.median(),
                'среднее': round(s.mean(), 2),
                'max': s.max(),
                'нулей_%': round((s == 0).mean() * 100, 2),
            })
        rows.append(row)
    return pd.DataFrame(rows).set_index('столбец')


profile_dataset(df)

,тип,пропусков_%,уникальных,мода,доля_моды_%,min,медиана,среднее,max,нулей_%
столбец,,,,,,,,,,
person_age,int64,0.0,56,23,11.91,20.00000,26.000000,27.73,9.400000e+01,0.00
person_income,int64,0.0,4294,60000,3.21,4000.00000,55000.000000,65908.58,2.039784e+06,0.00
person_home_ownership,category,0.0,4,RENT,50.53,NaN,NaN,NaN,NaN,NaN
person_emp_length,float64,0.0,35,0.0,12.61,0.00000,4.000000,4.76,4.100000e+01,12.61
loan_intent,category,0.0,6,EDUCATION,19.78,NaN,NaN,NaN,NaN,NaN
loan_grade,category,0.0,7,A,33.02,NaN,NaN,NaN,NaN,NaN
loan_amnt,int64,0.0,753,10000,8.17,500.00000,8000.000000,9593.99,3.500000e+04,0.00
loan_int_rate,float64,0.0,349,10.99,5.54,5.42000,10.990000,11.02,2.322000e+01,0.00
loan_status,int64,0.0,2,0,78.13,0.00000,0.000000,0.22,1.000000e+00,78.13


**Что стоит отметить:** `loan_status` — 78% нулей (дисбаланс классов); `person_home_ownership` почти целиком состоит из `RENT` и `MORTGAGE`, а `OTHER` встречается редко; `person_emp_length` содержит заметную долю нулей (стаж меньше года).

### 8.2. Доля просрочек по категориальным признакам

In [44]:
def default_rate_by(data: pd.DataFrame, col: str) -> pd.DataFrame:
    return (
        data.groupby(col, observed=True)
            .agg(n=(TARGET, 'size'), default_rate=(TARGET, 'mean'))
            .assign(share=lambda t: t['n'] / t['n'].sum())
            .sort_values('default_rate', ascending=False)
            .round(3)
    )


for col in CATEGORICAL_COLS + ['emp_length_missing']:
    display(Markdown(f'**{col}**'))
    display(default_rate_by(df, col))

**person_home_ownership**

,n,default_rate,share
person_home_ownership,,,
RENT,16378,0.316,0.505
OTHER,106,0.311,0.003
MORTGAGE,13368,0.126,0.412
OWN,2563,0.075,0.079


**loan_intent**

,n,default_rate,share
loan_intent,,,
DEBTCONSOLIDATION,5189,0.287,0.160
MEDICAL,6042,0.268,0.186
HOMEIMPROVEMENT,3594,0.262,0.111
PERSONAL,5497,0.199,0.170
EDUCATION,6411,0.173,0.198
VENTURE,5682,0.149,0.175


**loan_grade**

,n,default_rate,share
loan_grade,,,
G,64,0.984,0.002
F,241,0.705,0.007
E,963,0.645,0.030
D,3620,0.591,0.112
C,6437,0.208,0.199
B,10387,0.163,0.320
A,10703,0.100,0.330


**cb_person_default_on_file**

,n,default_rate,share
cb_person_default_on_file,,,
True,5730,0.379,0.177
False,26685,0.184,0.823


**emp_length_missing**

,n,default_rate,share
emp_length_missing,,,
True,889,0.317,0.027
False,31526,0.216,0.973


### 8.3. Связь числовых признаков с целевой переменной

Корреляция Пирсона с `loan_status` — грубая, но быстрая оценка: какие признаки линейно связаны с просрочкой.

In [45]:
df[NUMERIC_COLS].corrwith(df[TARGET]).sort_values(ascending=False).round(3).to_frame('corr с loan_status')

,corr с loan_status
loan_percent_income,0.380
loan_int_rate,0.334
loan_amnt,0.106
cb_person_cred_hist_length,-0.016
person_age,-0.022
person_emp_length,-0.086
person_income,-0.169


Сильнее всего с просрочкой связаны `loan_percent_income` и `loan_int_rate`. Корреляция ловит только линейную связь, поэтому посмотрим на долю просрочек по **квантильным корзинам** `loan_percent_income` (`pd.qcut` делит на группы примерно равного размера):

In [46]:
lpi_bins = pd.qcut(df['loan_percent_income'], q=5)
df.groupby(lpi_bins, observed=True).agg(
    n=(TARGET, 'size'),
    default_rate=(TARGET, 'mean'),
).round(3)

,n,default_rate
loan_percent_income,,
"(-0.001, 0.08]",7534,0.115
"(0.08, 0.12]",5558,0.125
"(0.12, 0.18]",7297,0.152
"(0.18, 0.25]",5776,0.192
"(0.25, 0.83]",6250,0.531


> **Гипотеза 2 (H2).** Связь долговой нагрузки с просрочкой **нелинейна**: в первых четырёх корзинах доля просрочек растёт плавно (11% → 19%), а когда кредит превышает примерно четверть годового дохода, она скачком выходит на уровень свыше 50%. Возможно, существует «порог нагрузки», после которого заёмщик перестаёт справляться с платежами.

### 8.4. Владение жильём: эффект сохраняется внутри грейда?

По таблице из 8.2 арендаторы (`RENT`) допускают просрочку заметно чаще. Но, может быть, у них просто хуже грейды? Проверим сводной таблицей: доля просрочек в разрезе «жильё × грейд».

In [47]:
pd.pivot_table(
    df, index='person_home_ownership', columns='loan_grade',
    values=TARGET, aggfunc='mean', observed=True,
).round(2)

loan_grade,A,B,C,D,E,F,G
person_home_ownership,,,,,,,
MORTGAGE,0.04,0.08,0.14,0.46,0.48,0.63,1.00
OTHER,0.12,0.15,0.35,0.55,0.75,1.00,NaN
OWN,0.07,0.04,0.06,0.06,0.54,0.53,1.00
RENT,0.18,0.24,0.27,0.74,0.76,0.78,0.96


In [48]:
pd.pivot_table(
    df, index='person_home_ownership', columns='loan_grade',
    values='person_income', aggfunc='median', observed=True,
).round(2)

loan_grade,A,B,C,D,E,F,G
person_home_ownership,,,,,,,
MORTGAGE,66000.0,70000.0,70000.0,68200.0,75000.0,79050.0,74000.0
OTHER,60800.0,60516.0,49000.0,61196.5,44250.0,72625.0,NaN
OWN,45000.0,48000.0,45000.0,48000.0,51000.0,42642.0,74000.0
RENT,49000.0,48000.0,46500.0,45247.5,50000.0,51600.0,49998.0


Внутри **каждого** грейда арендаторы допускают просрочку чаще, чем владельцы жилья и ипотечники (например, в грейде A — 18% против 4–7%). При этом медианный доход у `RENT` и `OWN` почти одинаков, а просрочек у `OWN` в разы меньше — значит, дело не только в доходе.

> **Гипотеза 3 (H3).** Статус владения жильём несёт информацию о риске, которую **не учитывают** ни кредитный грейд, ни доход: аренда — дополнительная регулярная нагрузка и признак меньшей финансовой устойчивости.

## 9. Итоги

| № | Гипотеза | Где найдена | Как проверить дальше |
|---|----------|-------------|----------------------|
| 1 | Пропуск стажа работы — самостоятельный сигнал риска | 4.2 | проверить важность флага `emp_length_missing` в модели |
| 2 | Связь нагрузки `loan_percent_income` с просрочкой нелинейна, есть порог около 0.25–0.3 | 8.3 | Разбить на более мелкие корзины и найти точку перелома; сравнить линейный и пороговый признак в модели |
| 3 | Владение жильём влияет на риск независимо от грейда и дохода | 8.4 | Сравнить модели с признаком и без него; проверить эффект внутри групп по доходу |

**Помните:** это гипотезы, а не выводы. Профилирование показывает **связи**, но не причины.

**Чек-лист предобработки, который стоит унести с собой:**

1. Перед загрузкой — заглянуть в файл (кодировка, разделитель, десятичный знак).
2. После загрузки — `shape`, `dtypes`, пропуски, дубликаты.
3. Привести типы и проверить, что преобразование не породило новых `NaN`.
4. Невалидные значения — это ошибки: заменить на `NaN` по правилам из предметной области.
5. Перед заполнением пропусков проверить, не информативен ли сам пропуск.
6. Медиана устойчивее среднего; по возможности — медиана **внутри осмысленной группы**.
7. Выбросы не удалять вслепую: удалять ошибки, реальные крайние значения — помечать флагом.
8. Сравнить данные «до» и «после».